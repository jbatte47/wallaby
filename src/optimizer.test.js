import {
  calculateDistance, getRoutePermutations, getRouteDistance, optimize, mapHopDistances,
} from './optimizer';

describe('optimizer', () => {
  describe('distance calculator', () => {
    it('should correctly calculate the distance between two spacial points', () => {
      const distance = calculateDistance({ x: 1, y: 1, z: 0 }, { x: 2, y: 1, z: 2 });
      expect(distance).toEqual(2.23606797749979);
    });
  });
  describe('destination permutations generator', () => {
    it('should create all possible permutations that begin and end with the origin', () => {
      const origin = { name: 'origin' };
      const destinations = [
        { name: 'a', meta: 'a stuff' },
        { name: 'b', meta: 'b is better' },
        { name: 'c', meta: 'c rulez' }
      ];
      const permutations = getRoutePermutations(origin, destinations);
      expect(permutations).toMatchSnapshot();
    });
  });
  describe('hop distance calculator', () => {
    it('should correctly calculate the distance for each hop', () => {
      const routeWithHopDistances = mapHopDistances([
        { x: 1, y: 1, z: 0 }, { x: 2, y: 1, z: 2 }, { x: 3, y: 2, z: 4 }
      ]);
      expect(routeWithHopDistances).toMatchSnapshot();
    });
  });
  describe('total route distance calculator', () => {
    it('should correctly calculate the total distance for a route', () => {
      const route = [{ distance: 0 }, { distance: 2.23 }, { distance: 2.44 }];
      const distance = getRouteDistance(route);
      expect(distance).toEqual(4.67);
    });
  });
  describe('create optimized delivery route', () => {
    it('should correctly generate an optimized route', () => {
      const origin = { name: 'origin', x: 0, y: 0, z: 0 };
      const destinations = [
        { name: 'a', meta: 'a stuff', x: 1, y: 1, z: 0 },
        { name: 'b', meta: 'b is better', x: 2, y: 1, z: 2 },
        { name: 'c', meta: 'c rulez', x: 3, y: 2, z: 4 }
      ];
      const optimizedRoute = optimize(origin, destinations);
      expect(optimizedRoute).toMatchSnapshot();
    });
  });
});
