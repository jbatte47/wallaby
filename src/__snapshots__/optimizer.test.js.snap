// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`optimizer create optimized delivery route should correctly generate an optimized route 1`] = `
Object {
  "distance": 11.446279000112114,
  "route": Array [
    Object {
      "distance": 0,
      "isOrigin": true,
      "name": "origin",
      "x": 0,
      "y": 0,
      "z": 0,
    },
    Object {
      "distance": 1.4142135623730951,
      "meta": "a stuff",
      "name": "a",
      "x": 1,
      "y": 1,
      "z": 0,
    },
    Object {
      "distance": 4.58257569495584,
      "meta": "c rulez",
      "name": "c",
      "x": 3,
      "y": 2,
      "z": 4,
    },
    Object {
      "distance": 2.449489742783178,
      "meta": "b is better",
      "name": "b",
      "x": 2,
      "y": 1,
      "z": 2,
    },
    Object {
      "distance": 3,
      "isOrigin": true,
      "name": "origin",
      "x": 0,
      "y": 0,
      "z": 0,
    },
  ],
}
`;

exports[`optimizer destination permutations generator should create all possible permutations that begin and end with the origin 1`] = `
Array [
  Array [
    Object {
      "isOrigin": true,
      "name": "origin",
    },
    Object {
      "meta": "a stuff",
      "name": "a",
    },
    Object {
      "meta": "b is better",
      "name": "b",
    },
    Object {
      "meta": "c rulez",
      "name": "c",
    },
    Object {
      "isOrigin": true,
      "name": "origin",
    },
  ],
  Array [
    Object {
      "isOrigin": true,
      "name": "origin",
    },
    Object {
      "meta": "b is better",
      "name": "b",
    },
    Object {
      "meta": "a stuff",
      "name": "a",
    },
    Object {
      "meta": "c rulez",
      "name": "c",
    },
    Object {
      "isOrigin": true,
      "name": "origin",
    },
  ],
  Array [
    Object {
      "isOrigin": true,
      "name": "origin",
    },
    Object {
      "meta": "c rulez",
      "name": "c",
    },
    Object {
      "meta": "a stuff",
      "name": "a",
    },
    Object {
      "meta": "b is better",
      "name": "b",
    },
    Object {
      "isOrigin": true,
      "name": "origin",
    },
  ],
  Array [
    Object {
      "isOrigin": true,
      "name": "origin",
    },
    Object {
      "meta": "a stuff",
      "name": "a",
    },
    Object {
      "meta": "c rulez",
      "name": "c",
    },
    Object {
      "meta": "b is better",
      "name": "b",
    },
    Object {
      "isOrigin": true,
      "name": "origin",
    },
  ],
  Array [
    Object {
      "isOrigin": true,
      "name": "origin",
    },
    Object {
      "meta": "b is better",
      "name": "b",
    },
    Object {
      "meta": "c rulez",
      "name": "c",
    },
    Object {
      "meta": "a stuff",
      "name": "a",
    },
    Object {
      "isOrigin": true,
      "name": "origin",
    },
  ],
  Array [
    Object {
      "isOrigin": true,
      "name": "origin",
    },
    Object {
      "meta": "c rulez",
      "name": "c",
    },
    Object {
      "meta": "b is better",
      "name": "b",
    },
    Object {
      "meta": "a stuff",
      "name": "a",
    },
    Object {
      "isOrigin": true,
      "name": "origin",
    },
  ],
]
`;

exports[`optimizer hop distance calculator should correctly calculate the distance for each hop 1`] = `
Array [
  Object {
    "distance": 0,
    "x": 1,
    "y": 1,
    "z": 0,
  },
  Object {
    "distance": 2.23606797749979,
    "x": 2,
    "y": 1,
    "z": 2,
  },
  Object {
    "distance": 2.449489742783178,
    "x": 3,
    "y": 2,
    "z": 4,
  },
]
`;
